### PR TITLE
AdSense/Doubleclick Fast Fetch unlayoutCallback experiment setup

### PR DESCRIPTION
**Pull request by keithwrightbos**
For AdSense/Doubleclick Fast Fetch, setup infrastructure to experiment with what state is cleared as part of unlayoutCallback.  Current behavior allows for AMP creatives to remain such that they remain visible to the user upon resumeCallback from SERP viewer.  Experiment with two new behaviors:

- clear all including AMP creatives.  This allows for measuring the uplift given by allowing AMP creatives to remain
- do not destroy ad promise for unrendered creatives (did not call layoutCallback as they did not get within 3 viewports).  This will allow for their usage should the user swipe back and scroll.

### CODE REVIEW

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js [hunk 1]
```diff
@@ -387,7 +394,26 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
...
+      case 'all':
+        // Ensure all creatives are removed.
+        this.isAmpCreative_ = false;
+        break;
+      case 'remain':
+        if (this.qqid_ && this.isAmpCreative_ === null) {
+          // Ad response received but not yet rendered.  Note that no fills
+          // would fall into this case even if layoutCallback has executed.
+          // Assume high probability of continued no fill therefore do not
+          // tear down.
```
Comment (jonkeller) [inline extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: Can you add more to the comment, explaining what happens if assumption is incorrect?

Comment (keithwrightbos) [reply extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: I think its clear that in that case we don't send an additional query that may fill causing some impression loss.

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js [hunk 2]
```diff
@@ -400,6 +426,8 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
     }
     this.ampAnalyticsConfig_ = null;
     this.qqid_ = null;
+    this.isAmpCreative_ = null;
+    return superResult;
```
Comment (bradfrizzell) [inline extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: Previously we were not returning anything from unlayoutcallback, and I'm not sure why we need the return valye now. Can you explain a little why we do?

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js [hunk 3]
```diff
@@ -387,7 +394,26 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
 
   /** @override */
   unlayoutCallback() {
-    super.unlayoutCallback();
+    switch (this.postAdResponseExperimentFeatures['unlayout_exp']) {
```
Comment (bradfrizzell) [inline extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: optional: enum for these experiment names to keep them consolidated / easily reference-able

File: extensions/amp-ad-network-doubleclick-impl/0.1/test/test-amp-ad-network-doubleclick-impl.js [hunk 4]
```diff
@@ -649,37 +649,98 @@ describes.realWin('amp-ad-network-doubleclick-impl', realWinConfig, env => {
...
+      }, element, impl, env);
+      element = setup[0];
+      impl = setup[1];
+      env = setup[2];
+      impl.buildCallback();
+      impl.win.ampAdSlotIdCounter = 1;
+      expect(impl.element.getAttribute('data-amp-slot-index')).to.not.be.ok;
+      impl.layoutMeasureExecuted_ = true;
+      impl.uiHandler = {applyUnlayoutUI: () => {}};
+      const placeholder = doc.createElement('div');
```
Comment (bradfrizzell) [inline extensions/amp-ad-network-doubleclick-impl/0.1/test/test-amp-ad-network-doubleclick-impl.js]: nit: use more descriptive name than placeholder, e.g. adSlotPlaceholder  or something like that

File: extensions/amp-ad-network-doubleclick-impl/0.1/test/test-amp-ad-network-doubleclick-impl.js [hunk 5]
```diff
@@ -649,37 +649,98 @@ describes.realWin('amp-ad-network-doubleclick-impl', realWinConfig, env => {
...
+    it('should not reset state to null on FIE unlayoutCallback', () => {
+      impl.onCreativeRender({customElementExtensions: []});
+      expect(impl.unlayoutCallback()).to.be.false;
+      expect(impl.iframe).is.ok;
+    });
+
+    it('should remove FIE if in all exp', () => {
+      impl.onCreativeRender({customElementExtensions: []});
+      impl.postAdResponseExperimentFeatures['unlayout_exp'] = 'all';
+      expect(impl.unlayoutCallback()).to.be.true;
```
Comment (bradfrizzell) [inline extensions/amp-ad-network-doubleclick-impl/0.1/test/test-amp-ad-network-doubleclick-impl.js]: optional: you do these 3 expects in several tests. could make a helper function

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js (hunk 2 above)
Comment (keithwrightbos) [reply extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: base-element#unlayoutCallback returns a boolean (https://github.com/ampproject/amphtml/blob/master/src/base-element.js#L487)

Comment (bradfrizzell) [reply extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: right, I was more curious as to why we didn't have anything being return before but now need to.

Comment (keithwrightbos) [reply extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: We should have been... I'm fixing it

Comment (bradfrizzell) [reply extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: gotcha, thanks. strange that the presubmits didn't catch that

**Pull request merged**